src/coord/src/coord.rs

```diff
@@ -3461,6 +3463,117 @@ pub async fn serve(
     }
 }
 
+pub fn serve_debug(
+    catalog_path: &Path,
+) -> (
+    JoinOnDropHandle<()>,
+    Client,
+    tokio::sync::mpsc::UnboundedSender<WorkerFeedbackWithMeta>,
+    tokio::sync::mpsc::UnboundedReceiver<WorkerFeedbackWithMeta>,
+    Arc<Mutex<u64>>,
+) {
+    lazy_static! {
+        static ref DEBUG_TIMESTAMP: Arc<Mutex<EpochMillis>> = Arc::new(Mutex::new(0));
+    }
+    pub fn get_debug_timestamp() -> EpochMillis {
+        *DEBUG_TIMESTAMP.lock().unwrap()
+    }
+
+    let (catalog, builtin_table_updates) = catalog::Catalog::open(&catalog::Config {
+        path: catalog_path,
+        enable_logging: true,
+        experimental_mode: None,
+        safe_mode: false,
+        cache_directory: None,
+        build_info: &DUMMY_BUILD_INFO,
+        num_workers: 0,
+        timestamp_frequency: Duration::from_millis(1),
+        now: get_debug_timestamp,
+    })
+    .unwrap();
+    let (cmd_tx, cmd_rx) = mpsc::unbounded_channel();
+    let (internal_cmd_tx, internal_cmd_rx) = mpsc::unbounded_channel();
+    let (ts_tx, ts_rx) = std::sync::mpsc::channel();
+    let (worker_tx, worker_rx) = crossbeam_channel::unbounded();
+    let worker_guards = dataflow::serve(dataflow::Config {
+        command_receivers: vec![worker_rx],
+        timely_worker: timely::WorkerConfig::default(),
+        experimental_mode: true,
+        now: get_debug_timestamp,
+    })
+    .unwrap();
+
+    // We want to be able to control communication from dataflow to the
+    // coordinator, so setup an additional channel pair.
+    let (feedback_tx, inner_feedback_rx) = mpsc::unbounded_channel();
+    let (inner_feedback_tx, feedback_rx) = mpsc::unbounded_channel();
+
+    // TODO: make the timestamper deterministic. It shouldn't run on its own
+    // internal 10ms loop.
+    let mut timestamper =
```

Comment:
if you're not working with kafka sources then the timestamper will be totally irrelevant anyway. we could change it to

- take a optional duration which if not specified turns it off
- give a transmitter that nothing else listens to instead of the `internal_cmd_tx` - then it will never be able to communicate to the coordinator anyway

Reply:
Thanks for the pushback. I've removed the use of a Timestamper and now have a simple thread that drains ts messages until it sees a shutdown. Determinism increased!